DM-28758: Create gen3 analysis_drp histogram plotting task

leeskelvin opened the pull request.

cmsaunders commented on python/lsst/analysis/drp/histPlot.py:
```diff
@@ -0,0 +1,387 @@
+            mask = np.isfinite(catPlot[column].values)
+            col = plt.cm.tab10(count)
+            rightHistAx.hist(catPlot[column][mask], bins=self.config.nBins, alpha=0.7, color=col,
+                             range=(rightHistLower, rightHistUpper), histtype="bar", lw=2)
```
You don't have labels or a legend for the right-hand histogram, because it is assumed that the data will correspond to the same measurements (i.e. PSFlux and 9ApFlux here), but could the user run into problems by setting different measurements? Also, if they list dictionary items differently in `leftHistColumns` and `rightHistColumns`, I think you would end up having the colors switched.

leeskelvin replied:
See main reply comment below which addresses the main issue regarding naming conventions/restrictions. With regards no label for the secondary y-axis, this is intentional due to the limited space on the horizontal axis. As both panels will always display a (number-count) histogram, the y-axes should always be some measure of counts/frequency, hence a unified y-axis naming convention here. 

It's not impossible to imagine switching one of the panels from displaying frequency/number histograms to density histograms instead, however, the code as-is currently isn't set up to construct such a plot, so this point is moot. If it's okay with you, I suggest putting off adding an additional y-axis label until such time as one is needed, preserving as much horizontal space as possible for figure plotting.

cmsaunders reviewed (commented):
This mostly looks good; there are a few minor comments below. My only major comment is that it seems like the same things need to be set in several places in the Config class, and then the Task makes assumptions that they are set the same way, for example having hist1 correspond to PSFlux and hist2 correspond to 9ApFlux in both `leftHistColumns`, `rightHistColumns`, and `setDefaults`. If you could set these in one place, then have that propagated, this would be more fool-proof. I am not sure whether that's possible, or in line with the rest of `analysis_drp`, but it's something to consider.

leeskelvin commented on python/lsst/analysis/drp/histPlot.py:
```diff
@@ -0,0 +1,387 @@
+    leftHistActions = ConfigurableActionStructField(
+        doc="The actions used to calculate the values for each histogram in the left panel.",
+        default={"hist1": SingleColumnAction,
+                 "hist2": SingleColumnAction, },
```
Removing the space after the comma raises the error: `missing whitespace after ','flake8(E231)`. I'm not attached to trailing commas however, so I'll remove it here (and everywhere else I've spotted it, e.g., lines 35, 55).

leeskelvin commented on python/lsst/analysis/drp/histPlot.py:
```diff
@@ -0,0 +1,387 @@
+    rightHistActions = ConfigurableActionStructField(
+        doc="The actions used to calculate the values for each histogram in the right panel.",
+        default={"hist1": calcFunctors.SNCalculator,
+                 "hist2": calcFunctors.SNCalculator, },
```
(comma removed instead)

leeskelvin commented on python/lsst/analysis/drp/histPlot.py:
```diff
@@ -0,0 +1,387 @@
+        catPlot : `pandas.core.frame.DataFrame`
+            The catalog to plot the points from.
+        dataId :
+        `lsst.daf.butler.core.dimensions._coordinate._ExpandedTupleDataCoordinate`
```
I wasn't sure about this either, and can't find anything in the numpydoc guidelines. I settled for this, as it mimics the behavior in `skyPlot.py` and `scatterPlot.py`. However, `plotUtils.py` splits this across two lines, which may be more appropriate?

cmsaunders replied:
The current version seems slightly more readable than splitting the line, so I am fine with keeping it.

leeskelvin commented on python/lsst/analysis/drp/histPlot.py:
```diff
@@ -0,0 +1,387 @@
+        leftHistAx.set_xlim(leftHistLower, leftHistUpper)
+        leftHistAx.set_xlabel(self.config.axisLabels["xLeft"])
+        leftHistAx.set_ylabel(self.config.axisLabels["yLeft"])
+        leftHistAx.legend(loc="lower right", bbox_to_anchor=(0.995, 0.1),
```
Also interesting! I tested this by changing `lower right` to other values, e.g., `upper right`. It looks like it does have an impact: the `loc` argument is setting the position of the anchor, whilst the `bbox_to_anchor` sets the coordinate of the anchor. Changing `lower right` to `upper right` has the effect here of moving the legend even further into the lower right corner. 

If I remove `loc` completely, it looks like it defaults to `upper right`. I think this is the wrong behavior here, as it means that the box will grow 'down' if more histograms are added. Setting `loc` to `lower right` means that the box will expand upwards into the empty white space if more data columns are requested.

leeskelvin commented:
Thanks for the review above Clare. I've pushed an update which I hope addresses most of the points above. With regards your main issue on mandating that certain things are named the same way across the code, I've been taking a closer look at over the last couple of days. I couldn't come up with a simple/clean means by which to set these names in a single place and use them everywhere else thereafter. I think adding that functionality would require a modifications within `lsst.pipe.tasks.configurableActions`. 

As a compromise, I've instead added additional checking in `HistPlotTaskConnections` which assigns dummy variables instead, should appropriate matches not be found. To accompany this, I've updated the names of these config variables to make more sense: `leftHistActions -> leftPanelActions`; `leftHistColumns -> leftPanelLabels` (and similarly for `right`). Now, if a matching dict key is not found in `leftPanelLabels`, a dummy name will be generated (e.g., `left_hist`). This label will be plotted on the figure, and used to select the data generated via the action.

Similarly, if the `summaryStatsLabel` (formerly `summaryStatsColumn`) cannot be matched in `leftPanelLabels` or `rightPanelLabels`, the first key in the dictionary is instead used to construct the marginal on-sky plot.

I hope these changes read well to you. If matches are not found, the code will no longer simply crash, instead producing a usable plot (albeit with unhelpful legend names). Keen to hear your thoughts on these changes, and the others made above.

leeskelvin commented on python/lsst/analysis/drp/histPlot.py:
```diff
@@ -0,0 +1,387 @@
+                                                [self.config.pLower, self.config.pUpper])
+            pLowers.append(pLower)
+            pUppers.append(pUpper)
+        leftHistLower, leftHistUpper = np.min(pLowers), np.max(pUppers)
```
I don't mind splitting it into two lines, if it helps readability, so I've done that here.

cmsaunders reviewed (approved):
Thanks for the changes. Everything looks good to me now.
